### PR TITLE
ZooReader/ZooReaderWriter cleanup

### PR DESCRIPTION
**Pull request by ctubbsii**
* Replace redundant retry loops throughout the ZooReader and
  ZooReaderWriter code with functional methods that can accept a lambda
* Simplify calls to ZooReader methods with null or unused Stat, or
  unused version parameters and remove unnecessary or redundant boolean
  return types
* Stop unnecessary handling of NONODE on delete (ZooReaderWriter already
  handles this internally)
* Remove unnecessary direct dependency on zookeeper-jute in minicluster
  pom.xml after removing unused ZK Stat object
* Move IllegalTableTransitionException to its own "outer" class file and
  make it extend RuntimeException to play more nicely with lambdas
* Remove unused or unnecessary overloaded methods in
  ZooReader/ZooReaderWriter

### CODE REVIEW

**Comment by ctubbsii** [comment]
All ITs passed (with once again, MetadataMaxFilesIT being the only flake: see #1574).

**Comment by milleruntime** [comment]
I took a quick look at this and I think the changes are solid but this PR is hard to review.  There is a lot of translating the code to lambdas combined with other changes.

**Comment by ctubbsii** [comment]
> I took a quick look at this and I think the changes are solid but this PR is hard to review. There is a lot of translating the code to lambdas combined with other changes.

Yeah, I know it's a little hard to follow. It didn't seem possible to make these improvements while avoiding that. Thanks for taking a look.

**Comment by keith-turner** [comment]
@ctubbsii this is a nice change. Removing all the passing of null is nice.  I noticed the mutation function is passed null,null in multiple places.  Was that something you looked into?

**Comment by ctubbsii** [comment]
> Was that something you looked into?

@keith-turner No, it wasn't. However, taking a quick look yielded PR #1758
